Mathlib/RingTheory/DedekindDomain/Factorization.lean

```diff
@@ -177,3 +223,289 @@ theorem finprod_heightOneSpectrum_factorization_coe (I : Ideal R) (hI : I ≠ 0)
+  push_cast
+  ring
+
+/-- For nonzero `I, I'`, `val_v(I*I') = val_v(I) + val_v(I')`. If `I` or `I'` is zero, then
+`val_v(I*I') = 0`. -/
+theorem count_mul' (I I' : FractionalIdeal (nonZeroDivisors R) K) :
+    count K v (I * I') = if I ≠ 0 ∧ I' ≠ 0 then count K v I + count K v I' else 0 := by
+  split_ifs with h
+  · exact count_mul K v h.1 h.2
+  · push_neg at h
+    by_cases hI : I = 0
+    · rw [hI, MulZeroClass.zero_mul, count, dif_pos (Eq.refl _)]
+    · rw [h hI, MulZeroClass.mul_zero, count, dif_pos (Eq.refl _)]
+
+/-- val_v(0) = 0. -/
+theorem count_zero : count K v (0 : FractionalIdeal (nonZeroDivisors R) K) = 0 := by
+  rw [count, dif_pos (Eq.refl _)]
+
+/-- val_v(1) = 0. -/
+theorem count_one : count K v (1 : FractionalIdeal (nonZeroDivisors R) K) = 0 := by
+  have h1 : (1 : FractionalIdeal (nonZeroDivisors R) K) =
+      spanSingleton (nonZeroDivisors R) ((algebraMap R K) 1)⁻¹ * ↑(1 : Ideal R) := by
+    rw [(algebraMap R K).map_one, Ideal.one_eq_top, coeIdeal_top, mul_one, inv_one,
+      spanSingleton_one]
+  rw [count_well_defined K v one_ne_zero h1, Ideal.span_singleton_one, Ideal.one_eq_top, sub_self]
+
+/-- For every `n ∈ ℕ` and every ideal `I`, `val_v(I^n) = n*val_v(I)`. -/
+theorem count_pow (n : ℕ) (I : FractionalIdeal (nonZeroDivisors R) K) :
+    count K v (I ^ n) = n * count K v I := by
+  induction' n with n h
+  · rw [pow_zero, ofNat_zero, MulZeroClass.zero_mul, count_one]
+  · rw [pow_succ, count_mul']
+    by_cases hI : I = 0
+    · have h_neg : ¬(I ≠ 0 ∧ I ^ n ≠ 0) := by
+        rw [not_and, Classical.not_not, Ne.def]
+        intro h
+        exact absurd hI h
+      rw [if_neg h_neg, hI, count_zero, MulZeroClass.mul_zero]
+    · rw [if_pos (And.intro hI (pow_ne_zero n hI)), h, Nat.succ_eq_add_one, Nat.cast_add,
+        Nat.cast_one]
+      ring
+
+/-- `val_v(v) = 1`, when `v` is regarded as a fractional ideal. -/
+theorem count_self : count K v (v.asIdeal : FractionalIdeal (nonZeroDivisors R) K) = 1 := by
+  have hv : (v.asIdeal : FractionalIdeal (nonZeroDivisors R) K) ≠ 0 := coeIdeal_ne_zero.mpr v.ne_bot
+  have h_self : (v.asIdeal : FractionalIdeal (nonZeroDivisors R) K) =
+      spanSingleton (nonZeroDivisors R) ((algebraMap R K) 1)⁻¹ * ↑v.asIdeal := by
+    rw [(algebraMap R K).map_one, inv_one, spanSingleton_one, one_mul]
+  have hv_irred : Irreducible (Associates.mk v.asIdeal) := by apply v.associates_irreducible
+  rw [count_well_defined K v hv h_self, Associates.count_self hv_irred, Ideal.span_singleton_one,
+    ← Ideal.one_eq_top, Associates.mk_one, Associates.factors_one, Associates.count_zero hv_irred,
+    ofNat_zero, sub_zero, ofNat_one]
+
+/-- `val_v(v) = n` for every `n ∈ ℕ`. -/
+theorem count_pow_self (n : ℕ) :
+    count K v ((v.asIdeal : FractionalIdeal (nonZeroDivisors R) K) ^ n) = n := by
+  rw [count_pow, count_self, mul_one]
+
+/-- `val_v(I⁻ⁿ) = -val_v(Iⁿ)` for every `n ∈ ℤ`. -/
+theorem count_inv (n : ℤ) (I : FractionalIdeal (nonZeroDivisors R) K) :
```

Comment:
I think `count_inv` should instead refer to `count K v (I⁻¹) = - count K v`. This maybe should be called `count_zpow_neg` or similar.

Sidenote: Do we need `count K v (I⁻¹) = - count K v`?

Reply:
I renamed this to `count_neg_zpow` and added `count_inv` for the special case.